cntp-trappstream/metainfofile.cpp

```diff
@@ -0,0 +1,172 @@
+#include "metainfofile.h"
+
+#include <QFile>
+#include <QXmlStreamReader>
+
+struct MetainfoFilePrivate {
+        QXmlStreamReader reader;
+        QXmlStreamWriter writer;
+
+        QString trContents;
+        QString trKey;
+        QString elementNamespace;
+        QString elementName;
+        QString preWhitespace;
+        QString leftWhitespace;
+        QString rightWhitespace;
+        QXmlStreamAttributes elementAttrs;
+
+        int translationLevel = -1;
+};
+
+MetainfoFile::MetainfoFile(QString file) {
+    d = new MetainfoFilePrivate();
+
+    QFile f(file);
+    f.open(QFile::ReadOnly);
+    d->reader.addData(f.readAll());
+    f.close();
+}
+
+MetainfoFile::~MetainfoFile() {
+    delete d;
+}
+
+void MetainfoFile::setOutput(QIODevice* device) {
+    d->writer.setDevice(device);
```

Comment:
It's a shame there isn't a "null device" we can use instead of checking `if (d->writer.device())` everywhere